Enhancements and Testing Summary

lily4178993 opened the pull request:
# Enhancements and Testing Summary

## Overview

In the **Space Travelers' Hub** project, our collaborative team consisting of @mryeminthu  and @lily4178993 has successfully developed a web application that empowers users to book rockets and participate in space missions.

## Components and Features

Our project revolves around three main sections: **Rockets**, **Missions**, and **My Profile**.

### Rockets

The Rockets section presents users with a curated list of available SpaceX rockets, offering the capability to reserve or cancel reservations. The user interface showcases essential rocket details, such as names, descriptions, and captivating images.

![Web capture Rockets](https://github.com/lily4178993/space-traveller/assets/101261047/c06228f7-a630-4dd2-931f-8b0083f53340)


### Missions

The Missions section provides users with a glimpse into the exciting world of space exploration. Users can explore current space missions, read engaging descriptions, and seamlessly join or leave missions, enhancing their immersive experience.

![Web capture Missions](https://github.com/lily4178993/space-traveller/assets/101261047/6b59779c-412b-44b6-af1c-065c013647e2)


### My Profile

The My Profile section tailors the user experience by offering a personalized overview of reserved rockets and joined missions. This intuitive interface empowers users to effortlessly track their participation history and journey into the cosmos.

![Web capture Profile](https://github.com/lily4178993/space-traveller/assets/101261047/55d1e408-9ef7-44cf-abac-3ed64d2f5afc)

## Implementation Highlights

Our team has demonstrated a dedicated focus on implementing key functionalities that ensure a seamless and captivating user experience:

- Real-time data retrieval from the SpaceX API for Rockets, Missions.
- Strategic utilization of Redux and Redux Toolkit to effectively manage and store fetched data, ensuring streamlined state management.
- Clever deployment of conditional rendering, dynamically displaying reserved and joined statuses for rockets, and missions.
- Seamless navigation facilitated by the integration of React Router, offering users an engaging and dynamic journey.

## Testing Efforts

Our commitment to delivering a reliable and stable application led us to invest substantial effort into testing. We conducted thorough unit and integration tests for critical sections, including navigation(**Header**), **MyProfile**, **Missions** and **Rockets**. This rigorous testing process encompassed simulating user interactions, validating data integrity, and ensuring smooth component rendering.

## Notes for Reviewers

Reviewers are encouraged to explore the application firsthand using the deployed [link](https://nelye-spacetravellershub.netlify.app/) or by setting up a local environment with the following commands:

```bash
git clone https://github.com/lily4178993/space-travellers.git
cd space-travellers
npm install
npx eslint "**/*.{js,jsx}" --fix
npm start
```

lily4178993 committed:
feat(initial): initialize Create React App project

lily4178993 committed:
feat(lint): add ESLint and Stylelint to project

lily4178993 committed:
fix(lint): resolve ESLint and Stylelint errors

lily4178993 committed:
chore(src): streamline project structure, keep only essential files

lily4178993 committed:
feat: Create Header component and set up Routers

lily4178993 committed:
feat: Add logo to Header component

lily4178993 committed:
feat: Fix css linter erors

mryeminthu committed:
Merge pull request #26 from lily4178993/feature/create-website-header

Create Header component and set up Routers

mryeminthu committed:
Merge pull request #27 from lily4178993/rocket-basic-structure

Basic rocket structure

lily4178993 committed:
feat(missions): implement mission page with NavLink navigation and data fetching

- Created a route and a view component for the Missions page.
- Utilized <NavLink /> for navigation links and styled the active class for user clarity.
- Established a directory structure for Redux state slice files.
- Implemented data fetching from the Missions endpoint (https://api.spacexdata.com/v3/missions).
- Dispatched an action to store mission data in Redux, including mission_id, mission_name, and description.

lily4178993 committed:
Merge pull request #28 from lily4178993/feature/missions-page-redux-setup

Mission Page Implementation and Data Fetching

Enhancements to the Mission Page have been merged into the dev branch. These improvements include navigation links, data fetching, and organized Redux state, enhancing the user interface and data management.

mryeminthu committed:
Add CSS to Header

mryeminthu committed:
Add CSS to header

mryeminthu committed:
Basic rocket structure

lily4178993 committed:
Merge branch 'feature/create-website-header' of https://github.com/lily4178993/space-traveller into feature/mission-page-enhancements

mryeminthu committed:
Merge pull request #31 from lily4178993/rocket-booking

Implement rocket booking

lily4178993 committed:
feat: Implement mission display, joining, and leaving

This commit includes the following changes:
1. Display missions in lists.
2. Implement mission joining with associated actions.
3. Implement mission leaving with associated actions.
4. Switch badges for missions using conditional components.
5. Persist join and leave actions to local storage.

lily4178993 committed:
Merge branch 'dev' into feature/mission-page-enhancements

lily4178993 committed:
feat(linter): fix js/jsx linter errors

lily4178993 committed:
Merge pull request #33 from lily4178993/feature/mission-page-enhancements

Mission Page Enhancements

Brings valuable enhancements to the Mission page, including mission display, join/leave actions, badge switching, and local storage persistence. These improvements enhance the overall user experience and functionality of the Mission page.

mryeminthu committed:
Implement rocket booking

mryeminthu committed:
Merge pull request #34 from lily4178993/rocket-in-profile

Create MyProfile.jsx and profile.css

lily4178993 committed:
fix: Resolve merge conflict in App.js

lily4178993 committed:
chore: Organize mission component styles in missions.css

mryeminthu committed:
Remove one Reserved button

lily4178993 committed:
feat(profile): Render list of joined missions on 'My profile' page and remove local storage data

lily4178993 committed:
feat(linter): fix js/jsx linter errors

mryeminthu committed:
Add CSS to Header

mryeminthu committed:
Create MyProfile.jsx and profile.css

mryeminthu committed:
USe conditional statement in Rocket.jsx

lily4178993 committed:
Merge pull request #35 from lily4178993/feature/render-joined-missions

Feature: Render List of Joined Missions on My Profile Page

- Added rendering of joined missions on the "My profile" page using the `filter()` method.
- Removed the code related to persisting data in local storage.

lily4178993 committed:
feat(testing): add Jest as a development dependency

mryeminthu committed:
Update README.md

mryeminthu committed:
Update README.md

lily4178993 committed:
test: add unit tests for Header Component

mryeminthu committed:
Merge pull request #37 from lily4178993/unit-tests/header-component

Add Unit Tests for Header Component

mryeminthu committed:
Merge pull request #36 from lily4178993/readme

Add Readme File

mryeminthu committed:
Add css to rockets.css

lily4178993 committed:
test(MyProfile): add unit tests for rendering and data display

Added unit tests to ensure that the MyProfile component correctly renders the My Missions and My Rockets sections. Tests cover scenarios with and without data, and all tests passed successfully.

lily4178993 committed:
Merge pull request #39 from lily4178993/unit-tests/profile--component

Add unit tests for rendering and data display for MyProfile components

lily4178993 committed:
test(Rockets Component): should render rocket cards

This test ensures that the Rockets component correctly renders rocket cards based on the provided mock store state.

lily4178993 committed:
test(mission): add test for rendering mission data

lily4178993 committed:
Merge branch 'dev' into add-css

lily4178993 committed:
Merge pull request #38 from lily4178993/add-css

Add CSS styles to the navigation

mryeminthu committed:
Change rockets.css

mryeminthu committed:
Update Readme

mryeminthu committed:
Delete old readme

mryeminthu committed:
Update readme

mryeminthu committed:
Get old readme

mryeminthu committed:
Merge branch 'readme' of https://github.com/lily4178993/space-traveller into readme

mryeminthu committed:
Update readme

mryeminthu committed:
Modify profile css

mryeminthu committed:
Merge pull request #42 from lily4178993/unit-tests/added-tests-components

test(summury): Add comprehensive test suite for project

mryeminthu committed:
Add css to navbar, rocket and profile

mryeminthu committed:
Fix linter

mryeminthu committed:
Update readme

lily4178993 committed:
Revert "test(summury): Add comprehensive test suite for project"

lily4178993 committed:
Merge branch 'revert-42-unit-tests/added-tests-components' into unit-tests/missions-component

lily4178993 committed:
Merge pull request #41 from lily4178993/unit-tests/missions-component

Add mission data rendering test

lily4178993 committed:
Merge pull request #40 from lily4178993/unit-tests/rockets-component

Add test for rendering rocket cards in Rockets component

lily4178993 committed:
Merge pull request #43 from lily4178993/revert-42-unit-tests/added-tests-components

Revert "test(summury): Add comprehensive test suite for project"

lily4178993 committed:
feat(test): Addeed missing tests

lily4178993 committed:
Merge pull request #44 from lily4178993/fix-missing-tests

Added missing tests

lily4178993 committed:
feat: remove unecessary file

DammyShittu commented on src/components/Rockets.jsx:
```diff
@@ -0,0 +1,69 @@
+  const rocketData = useSelector((state) => state.rockets);
+
+  useEffect(() => {
+    if (rocketData.length === 0) {
+      fetch('https://api.spacexdata.com/v3/rockets')
+        .then((response) => response.json())
+        .then((data) => {
+          dispatch(fetchRockets(data));
+        })
```
- [ ] Kindly filter the rocket data from the API before dispatching it to the store, based on [this requirement](https://github.com/microverseinc/curriculum-react-redux/blob/main/group-project/project_space_travelers_hub.md#redux-fetch-data-and-update-redux-store).

  ![image](https://github.com/lily4178993/space-traveller/assets/54848363/c6728052-0d46-41b5-84da-bbaf18832cd6)

DammyShittu commented on src/redux/missions/missionsSlice.js:
```diff
@@ -0,0 +1,105 @@
+ */
+const fetchMissions = createAsyncThunk('missions/fetchMissions', async () => {
+  const response = await axios.get(MISSION_API_URL);
+  return response.data;
```
- [ ] Kindly filter the data from the missions API too. It is not all the information from the API that should be dispatched to the store.

  ![image](https://github.com/lily4178993/space-traveller/assets/54848363/358a9c29-06df-4513-8b09-ad87e3012427)

DammyShittu commented on package.json:
```diff
@@ -0,0 +1,62 @@
+    "react-redux": "^8.1.3",
+    "react-router-dom": "^6.16.0",
+    "react-scripts": "5.0.1",
+    "react-test-renderer": "^18.2.0",
+    "redux-mock-store": "^1.5.4",
+    "web-vitals": "^2.1.4"
+  },
+  "scripts": {
+    "start": "react-scripts start",
+    "build": "react-scripts build",
+    "test": "react-scripts test --transformIgnorePatterns \"node_modules/(?!axios)/\"",
+    "eject": "react-scripts eject"
+  },
+  "eslintConfig": {
+    "extends": [
+      "react-app",
+      "react-app/jest"
+    ]
+  },
+  "browserslist": {
+    "production": [
+      ">0.2%",
+      "not dead",
+      "not op_mini all"
+    ],
+    "development": [
+      "last 1 chrome version",
+      "last 1 firefox version",
+      "last 1 safari version"
+    ]
+  },
+  "devDependencies": {
+    "@babel/core": "^7.23.0",
+    "@babel/eslint-parser": "^7.22.15",
+    "@babel/plugin-syntax-jsx": "^7.22.5",
+    "@babel/preset-react": "^7.22.15",
+    "eslint": "^7.32.0",
+    "eslint-config-airbnb": "^18.2.1",
+    "eslint-plugin-import": "^2.28.1",
+    "eslint-plugin-jsx-a11y": "^6.7.1",
```
- [ ] From the requirement, the redux logger should be present in the `package.json` file. Kindly add it. When you add it and configure the store with it, you can view the store in the browser console.

  ```
  npm install -S redux-logger
  ```

DammyShittu commented on src/components/Header.jsx:
```diff
@@ -0,0 +1,21 @@
+    </div>
+    <nav>
+      <ul>
+        <li><Link to="/">Rockets</Link></li>
+        <li><Link to="/missions">Missions</Link></li>
+        <li><Link to="/profile">My Profile</Link></li>
```
- [ ] From the requirement, you should use <NavLink>, not <Link> for your header links. NavLink has the benefit of allowing us style the selected menu option, which is also a requirement.

DammyShittu reviewed (changes requested):
Hi @lily4178993 and @mryeminthu,

## Highlights
✔️ API is fetched once

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

lily4178993 committed:
feat(missions): refactor fetchMissions to fetch specific fields

Refactor the fetchMissions async thunk to fetch only mission_id, description,
mission_name, and set reserved to false initially.

lily4178993 committed:
feat(redux): add Redux Logger middleware to store configuration

lily4178993 committed:
Merge pull request #46 from lily4178993/fix/mission-slice-data-fetching

This merge incorporates changes from two branches: fix/mission-slice-data-fetching and redux-logger-integration. It optimizes mission data fetching to retrieve essential fields, including mission_id, description, mission_name, and reserved. Additionally, Redux Logger has been integrated to improve debugging and monitoring of the Redux store.

lily4178993 committed:
feat: Add active link highlighting using NavLink

Updated the Header component to use NavLink from 'react-router-dom' for navigation links.
Added the 'activeClassName' property to NavLink components to apply the 'active-link' CSS class when a link is active, providing visual highlighting for the active route.

lily4178993 committed:
feat: fix CSS linter errors

lily4178993 committed:
Merge pull request #47 from lily4178993/fix/navigation

Merge branch 'fix/navigation-active-link' into dev

lily4178993 committed:
feat: Improve rocket data fetching and handling

This commit enhances the rocket data fetching process from the SpaceX API. It also improves the way data is filtered and dispatched to the Redux store, ensuring that only the necessary fields (description, rocket_name, rocket_id, flickr_images, and reserved) are included.

Additionally, the bug related to the handle functions not working after the update has been resolved. The handleReserve and handleCancelReservation functions now function correctly.

lily4178993 committed:
Merge pull request #48 from lily4178993/fix/rocket-data-filtering

Merged: Improved Rocket Data Fetching and Bug Fix

This merge incorporates enhancements to the Rockets component and a critical bug fix. The rocket data fetching process has been optimized to fetch only necessary fields, and the `handleReserve` and `handleCancelReservation` functions have been fixed.

These changes improve the overall efficiency and reliability of the Rockets component.

lily4178993 commented:
Thank you @DammyShittu for your thorough inspection and valuable suggestions. Your keen eye for detail and commitment to quality have greatly contributed to the improvement of this project. We appreciate your dedication to excellence!

uwadonat reviewed (approved):
## Hi @Team,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
### Congratulations! 🎉 
### Highlights
- Detailed pull request message ✔️
- The application work as expected ✔️
- No liner errors ✔️
![apprrrr](https://github.com/lily4178993/space-traveller/assets/65860678/d8b28a59-7d4f-4ad5-8e48-013527feb9fe)

# Status: Approved ✔️✔️✔️

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100%.
Please, remember to tag me @uwadonat in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

lily4178993 commented:
We want to express our sincere gratitude to the reviewers for their careful evaluation and approval of this project. Your feedback and guidance have been invaluable in ensuring the success of this endeavor. Thank you for your time, insights, and support throughout this journey!
